codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,299 @@
+ * the License.
+ */
+
+package io.netty.handler.codec.redis;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.ByteProcessor;
+import io.netty.util.CharsetUtil;
+import io.netty.util.internal.PlatformDependent;
+
+import java.util.List;
+
+/**
+ * Decodes the Redis protocol into {@link RedisMessage} objects following
+ * <a href="http://redis.io/topics/protocol">RESP (REdis Serialization Protocol)</a>.
+ *
+ * {@link RedisMessage} parts can be aggregated to {@link RedisMessage} using
+ * {@link RedisArrayAggregator} or processed directly.
+ */
+public class RedisDecoder extends ByteToMessageDecoder {
+
+    private static final int CRLF_LENGTH = 2;
+    private static final int NULL_RESPONSE_LENGTH = -1;
+    private static final int REDIS_MESSAGE_MAX_LENGTH = 512 * 1024 * 1024; // 512MB
+    private static final int POSITIVE_LONG_MAX_LENGTH = 19; // length of Long.MAX_VALUE
+    private static final short EOL = PlatformDependent.BIG_ENDIAN_NATIVE_ORDER
+            ? (short) (((int) '\n' << 8) | '\r')
+            : (short) (((int) '\r' << 8) | '\n');
+
+    private final ToPositiveLongProcessor toPositiveLongProcessor = new ToPositiveLongProcessor();
+
+    private int maxInlineMessageLength;
+    private int maxBulkStringChunkSize;
+
+    // current decoding states
+    private State state = State.DECODE_TYPE;
+    private RedisMessageType type;
+    private int expectedBulkLength;
+    private int toReadBulkLength;
+
+    private enum State {
+        DECODE_TYPE,
+        DECODE_INLINE, // SIMPLE_STRING, ERROR, INTEGER
+        DECODE_LENGTH, // BULK_STRING, ARRAY_HEADER
+        DECODE_BULK_STRING,
+    }
+
+    /**
+     * Creates a new instance with default {@code maxInlineMessageLength (1024 * 64)} and
+     * default {@code maxBulkStringChunkSize (128KB)}.
+     */
+    public RedisDecoder() {
+        this(128 * 1024); // default chunk size.
+    }
+
+    /**
+     * Creates a new instance with default {@code maxInlineMessageLength (1024 * 64)}.
+     *
+     * @param maxBulkStringChunkSize the maximum length of bulk string chunk length.
+     */
+    public RedisDecoder(int maxBulkStringChunkSize) {
+        // 1024 * 64 is max inline length of current Redis server implementation.
+        this(maxBulkStringChunkSize, 1024 * 64);
+    }
+
+    /**
+     * Creates a new instance.
+     * @param maxBulkStringChunkSize the maximum length of bulk string chunk length.
+     * @param maxInlineMessageLength the maximum length of inline message.
+     */
+    public RedisDecoder(int maxBulkStringChunkSize, int maxInlineMessageLength) {
+        if (maxBulkStringChunkSize <= 0 || maxBulkStringChunkSize > REDIS_MESSAGE_MAX_LENGTH) {
+            throw new RedisCodecException("maxBulkStringChunkSize: " + maxBulkStringChunkSize +
+                                          " (expected: <= " + REDIS_MESSAGE_MAX_LENGTH + ")");
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal " (expected: <= " 4 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1192)